sdv/datasets/demo.py

```diff
@@ -39,41 +39,126 @@ def _validate_output_folder(output_folder_name):
         )
 
 
+def _create_s3_client():
+    """Create and return an S3 client with unsigned requests."""
+    return boto3.client('s3', config=Config(signature_version=SIGNATURE_VERSION))
+
+
 def _get_data_from_bucket(object_key):
-    session = boto3.Session()
-    s3 = session.client('s3', config=Config(signature_version=SIGNATURE_VERSION))
+    s3 = _create_s3_client()
     response = s3.get_object(Bucket=BUCKET, Key=object_key)
     return response['Body'].read()
 
 
+def _list_objects(prefix):
+    """List all objects under a given prefix using pagination.
+
+    Args:
+        prefix (str):
+            The S3 prefix to list.
+
+    Returns:
+        list[dict]:
+            A list of object summaries.
+    """
+    client = _create_s3_client()
+    contents = []
+    paginator = client.get_paginator('list_objects_v2')
+    for resp in paginator.paginate(Bucket=BUCKET, Prefix=prefix):
+        contents.extend(resp.get('Contents', []))
+
+    if not contents:
+        raise DemoResourceNotFoundError(f"No objects found under '{prefix}' in bucket '{BUCKET}'.")
+
+    return contents
+
+
+def _find_data_zip_key(contents, dataset_prefix):
+    """Find the 'data.zip' object key under dataset prefix, case-insensitive.
+
+    Args:
+        contents (list[dict]):
+            List of objects from S3.
+        dataset_prefix (str):
+            Prefix like 'single_table/dataset/'.
+
+    Returns:
+        str:
+            The key to the data zip if found.
+    """
+    prefix_lower = dataset_prefix.lower()
+    for entry in contents:
+        key = entry.get('Key', '')
+        key_lower = key.lower()
+        if key_lower == f'{prefix_lower}data.zip':
+            return key
```

Comment:
Seems like we do something similar to this in 2 other places, `_get_first_v1_metadata_bytes` and `iter_metainfo_yaml_entries`. Should we create one helper function that searches the contents for a given pattern and returns a list of all matches and use that instead?